File: src/main/java/com/example/hypocaust/models/enums/ModelSpecEnum.java
package com.example.hypocaust.models.enums;

import com.example.hypocaust.exception.ModelException;
import java.util.Arrays;

/**
 * Common interface for model specification enums that need JsonCreator functionality. Provides a
 * generic fromString method to eliminate code duplication across model enums.
 */
public interface ModelSpecEnum {

  String getModelName();

  /**
   * Generic fromString method for model enums that eliminates boilerplate code.
   *
   * @param modelName the model name string from configuration
   * @param enumClass the enum class to search within
   * @param modelType descriptive name for error messages
   * @param <T> enum type that extends both Enum and ModelSpecEnum
   * @return the matching enum constant
   * @throws ModelException if the model name is not recognized
   */
  static <T extends Enum<T> & ModelSpecEnum> T fromString(
      String modelName,
      Class<T> enumClass,
      String modelType) throws ModelException {
    return Arrays.stream(enumClass.getEnumConstants())
        .filter(model -> model.getModelName().equals(modelName))
        .findFirst()
        .orElseThrow(() -> new ModelException("Unknown " + modelType + " model: " + modelName));
  }
}

File: src/main/kotlin/com/example/the_machine/models/enums/ModelSpecEnum.kt
package com.example.the_machine.models.enums

import com.example.the_machine.exception.ModelException

/**
 * Common interface for model specification enums that need JsonCreator functionality. Provides a
 * generic fromString method to eliminate code duplication across model enums.
 */
interface ModelSpecEnum {

  val modelName: String

  companion object {

    /**
     * Generic fromString method for model enums that eliminates boilerplate code.
     * Uses Kotlin's reified type parameters and inline functions for improved performance.
     *
     * @param modelName the model name string from configuration
     * @param modelType descriptive name for error messages
     * @param T enum type that implements ModelSpecEnum
     * @return the matching enum constant
     * @throws ModelException if the model name is not recognized
     */
    inline fun <reified T> fromString(
      modelName: String,
      modelType: String
    ): T where T : Enum<T>, T : ModelSpecEnum {
      return enumValues<T>()
        .find { it.modelName == modelName }
        ?: throw ModelException("Unknown $modelType model: $modelName")
    }
  }
}